### PR TITLE
Add RichTextBlock support to Block Kit Kotlin DSL builder

### PR DESCRIPTION
**Pull request by KENNYSOFT**
This PR adds RichTextBlock support to Block Kit Kotlin DSL builder. With this changes, we can now use DSL like:

```kotlin
withBlocks {
    richText {
        richTextSection {
            text("Check out these different block types with paragraph breaks between them:\n\n")
        }
        richTextPreformatted {
            text("Hello there, I am preformatted block!\n\nI can have multiple paragraph breaks within the block.")
        }
        richTextSection {
            text("\nI am rich text with a paragraph break following preformatted text. \n\nI can have multiple paragraph breaks within the block.\n\n")
        }
        richTextQuote {
            text("I am a basic rich text quote, \n\nI can have multiple paragraph breaks within the block.")
        }
        richTextSection {
            text("\nI am rich text with a paragraph after the quote block\n\n")
        }
        richTextQuote {
            text("I am a basic quote block following rich text")
        }
        richTextSection {
            text("\n")
        }
        richTextPreformatted {
            text("I am more preformatted text following a quote block")
        }
        richTextSection {
            text("\n")
        }
        richTextQuote {
            text("I am a basic quote block following preformatted text")
        }
        richTextSection {
            text("\n")
        }
        richTextList {
            style(ListStyle.BULLET)
            richTextSection {
                text("list item one")
            }
            richTextSection {
                text("list item two")
            }
        }
        richTextSection {
            text("\nI am rich text with a paragraph break after a list")
        }
    }
}
```

### Category (place an `x` in each of the `[ ]`)

* [ ] **bolt** (Bolt for Java)
* [ ] **bolt-{sub modules}** (Bolt for Java - optional modules)
* [ ] **slack-api-client** (Slack API Clients)
* [x] **slack-api-model** (Slack API Data Models)
* [x] **slack-api-*-kotlin-extension** (Kotlin Extensions for Slack API Clients)
* [ ] **slack-app-backend** (The primitive layer of Bolt for Java)

## Requirements

Please read the [Contributing guidelines](https://github.com/slackapi/java-slack-sdk/blob/main/.github/contributing.md) and [Code of Conduct](https://slackhq.github.io/code-of-conduct) before creating this issue or pull request. By submitting, you agree to those rules.

### CODE REVIEW

File: slack-api-client/src/main/java/com/slack/api/util/json/GsonFactory.java [hunk 1]
```diff
@@ -87,7 +87,6 @@ public static Gson createCamelCase(SlackConfig config) {
                 .registerTypeAdapter(ContextBlockElement.class, new GsonContextBlockElementFactory(failOnUnknownProps))
                 .registerTypeAdapter(BlockElement.class, new GsonBlockElementFactory(failOnUnknownProps))
                 .registerTypeAdapter(RichTextElement.class, new GsonRichTextElementFactory(failOnUnknownProps))
-                .registerTypeAdapter(MessageChangedEvent.PreviousMessage.class, new GsonMessageChangedEventPreviousMessageFactory(failOnUnknownProps))
```
Comment (seratch) [inline slack-api-client/src/main/java/com/slack/api/util/json/GsonFactory.java:90]: Can you revert this change?

File: slack-api-model-kotlin-extension/src/main/kotlin/com/slack/api/model/kotlin_extension/block/element/ButtonStyle.kt [hunk 2]
```diff
@@ -6,16 +6,16 @@ enum class ButtonStyle {
...
      * Danger gives buttons a red outline and text, and should be used when the action is destructive. Use danger even
      * more sparingly than primary.
      */
     DANGER {
-        override val value: String? = "danger"
+        override val value = "danger"
     };
 
-    abstract val value: String?
+    abstract val value: String
```
Comment (seratch) [inline slack-api-model-kotlin-extension/src/main/kotlin/com/slack/api/model/kotlin_extension/block/element/ButtonStyle.kt:20]: Good catch

File: slack-api-model-kotlin-extension/src/main/kotlin/com/slack/api/model/kotlin_extension/block/element/container/SingleBlockElementContainer.kt [hunk 3]
```diff
@@ -9,7 +9,7 @@ import com.slack.api.model.kotlin_extension.block.element.dsl.BlockElementDsl
 /**
  * Supports a BlockElementContainer holding exactly one block element
  */
-class SingleBlockElementContainer() : BlockElementDsl {
```
Comment (seratch) [inline slack-api-model-kotlin-extension/src/main/kotlin/com/slack/api/model/kotlin_extension/block/element/container/SingleBlockElementContainer.kt:12]: I see this could be slightly better, but can you remove unrelated changes like this from this PR?

File: slack-api-model-kotlin-extension/src/main/kotlin/com/slack/api/model/kotlin_extension/block/composition/dsl/RichTextObjectDsl.kt [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+     * @param fallback Text to display in place of the date should parsing, formatting or displaying fail.
+     * @see <a href="https://api.slack.com/reference/block-kit/blocks#date-element-type">Rich text object documentation</a>
+     */
+    fun date(timestamp: Int, format: String, url: String? = null, fallback: String? = null)
+
+    /**
+     * @param name The name of the emoji; i.e. "wave" or "wave::skin-tone-2".
+     * @see <a href="https://api.slack.com/reference/block-kit/blocks#emoji-element-type">Rich text object documentation</a>
+     */
+    fun emoji(name: String)
```
Comment (seratch) [inline slack-api-model-kotlin-extension/src/main/kotlin/com/slack/api/model/kotlin_extension/block/composition/dsl/RichTextObjectDsl.kt:43]: emoji method should accept two patterns:
- name: String, skinTone: Int? = null, style: TextStyle? = null
- unicode: String

overloading with string type may not work; if that's the case, skipping unicode here is okay

File: slack-api-model-kotlin-extension/src/main/kotlin/com/slack/api/model/kotlin_extension/block/composition/dsl/RichTextObjectDsl.kt [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+import com.slack.api.model.kotlin_extension.block.element.BroadcastRange
+
+@BlockLayoutBuilder
+interface RichTextObjectDsl {
+
+    /**
+     * @param range The range of the broadcast.
+     * @see <a href="https://api.slack.com/reference/block-kit/blocks#broadcast-element-type">Rich text object documentation</a>
+     */
+    fun broadcast(range: BroadcastRange)
```
Comment (seratch) [inline slack-api-model-kotlin-extension/src/main/kotlin/com/slack/api/model/kotlin_extension/block/composition/dsl/RichTextObjectDsl.kt:15]: style: LimitedTextStyle? = null should be accepted too

File: slack-api-model-kotlin-extension/src/main/kotlin/com/slack/api/model/kotlin_extension/block/composition/dsl/RichTextObjectDsl.kt [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+    /**
+     * @param value The hex value for the color.
+     * @see <a href="https://api.slack.com/reference/block-kit/blocks#color-element-type">Rich text object documentation</a>
+     */
+    fun color(value: String)
```
Comment (seratch) [inline slack-api-model-kotlin-extension/src/main/kotlin/com/slack/api/model/kotlin_extension/block/composition/dsl/RichTextObjectDsl.kt:21]: same as above

File: slack-api-model-kotlin-extension/src/main/kotlin/com/slack/api/model/kotlin_extension/block/composition/dsl/RichTextObjectDsl.kt [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    fun channel(channelId: String, style: LimitedTextStyle? = null)
+
+    /**
+     * @param timestamp A Unix timestamp for the date to be displayed in seconds.
+     * @param format A template string containing curly-brace-enclosed tokens to substitute your provided timestamp.
+     * @param url URL to link the entire format string to.
+     * @param fallback Text to display in place of the date should parsing, formatting or displaying fail.
+     * @see <a href="https://api.slack.com/reference/block-kit/blocks#date-element-type">Rich text object documentation</a>
+     */
+    fun date(timestamp: Int, format: String, url: String? = null, fallback: String? = null)
```
Comment (seratch) [inline slack-api-model-kotlin-extension/src/main/kotlin/com/slack/api/model/kotlin_extension/block/composition/dsl/RichTextObjectDsl.kt:37]: same as above

File: slack-api-model/src/main/java/com/slack/api/model/block/element/RichTextQuoteElement.java [hunk 8]
```diff
@@ -19,5 +19,6 @@ public class RichTextQuoteElement extends BlockElement implements RichTextElemen
     private final String type = TYPE;
     @Builder.Default
     private List<RichTextElement> elements = new ArrayList<>();
+    private Integer border;
```
Comment (seratch) [inline slack-api-model/src/main/java/com/slack/api/model/block/element/RichTextQuoteElement.java:22]: Great catch; to make the automated tests pass, you need to run `./scripts/run_no_prep_tests.sh` to generate corresponding JSON files, which are used by other projects.

**Review by seratch — CHANGES_REQUESTED** [review]
Great work! Overall, your changes look great to me but could you take a bit more time to adjust the following?
- A few more adjustments to your changes; see my review comments
- Run `./scripts/run_no_prep_tests.sh` to generate corresponding JSON files
- Sign our CLA; without it, we are unable to accept any external contribution

Thank you!

File: slack-api-client/src/main/java/com/slack/api/util/json/GsonFactory.java (hunk 1 above)
Comment (KENNYSOFT) [reply slack-api-client/src/main/java/com/slack/api/util/json/GsonFactory.java:90]: Yes. Although this line is duplicated, I'll make a different PR for the code enhancements.

File: slack-api-model-kotlin-extension/src/main/kotlin/com/slack/api/model/kotlin_extension/block/composition/dsl/RichTextObjectDsl.kt (hunk 4 above)
Comment (KENNYSOFT) [reply slack-api-model-kotlin-extension/src/main/kotlin/com/slack/api/model/kotlin_extension/block/composition/dsl/RichTextObjectDsl.kt:43]: https://api.slack.com/reference/block-kit/blocks#emoji-element-type

image

I've implemented based on the documents, but yes I've seen that `RichTextSectionElement.Emoji` has those fields and will take a look for them.

File: slack-api-model/src/main/java/com/slack/api/model/block/element/RichTextQuoteElement.java (hunk 8 above)
Comment (KENNYSOFT) [reply slack-api-model/src/main/java/com/slack/api/model/block/element/RichTextQuoteElement.java:22]: Thanks for the guide! Will proceed.

**Comment by KENNYSOFT** [comment]
> * Sign our CLA; without it, we are unable to accept any external contribution

@seratch About CLA - I'm experiencing the same issue of https://github.com/slackapi/java-slack-sdk/pull/1364#issuecomment-2338727587. Can you check it?

image

**Comment by KENNYSOFT** [comment]
Thanks!

**Comment by seratch** [comment]
As a workaround, I closed and reopend this PR to let the CLA bot check it again. It seems your status is now correctly updated!

File: slack-api-client/src/main/java/com/slack/api/util/json/GsonFactory.java (hunk 1 above)
Comment (KENNYSOFT) [reply slack-api-client/src/main/java/com/slack/api/util/json/GsonFactory.java]: Resolved in 1be897c8ce41fb2d32ee1b4cdaa0d647fd7b43c0

File: slack-api-model-kotlin-extension/src/main/kotlin/com/slack/api/model/kotlin_extension/block/element/ButtonStyle.kt (hunk 2 above)
Comment (KENNYSOFT) [reply slack-api-model-kotlin-extension/src/main/kotlin/com/slack/api/model/kotlin_extension/block/element/ButtonStyle.kt]: But also reverted in 1be897c8ce41fb2d32ee1b4cdaa0d647fd7b43c0 😅

File: slack-api-model-kotlin-extension/src/main/kotlin/com/slack/api/model/kotlin_extension/block/element/container/SingleBlockElementContainer.kt (hunk 3 above)
Comment (KENNYSOFT) [reply slack-api-model-kotlin-extension/src/main/kotlin/com/slack/api/model/kotlin_extension/block/element/container/SingleBlockElementContainer.kt]: Resolved in 1be897c8ce41fb2d32ee1b4cdaa0d647fd7b43c0

File: slack-api-model-kotlin-extension/src/main/kotlin/com/slack/api/model/kotlin_extension/block/composition/dsl/RichTextObjectDsl.kt (hunk 4 above)
Comment (KENNYSOFT) [reply slack-api-model-kotlin-extension/src/main/kotlin/com/slack/api/model/kotlin_extension/block/composition/dsl/RichTextObjectDsl.kt]: Resolved in 1be897c8ce41fb2d32ee1b4cdaa0d647fd7b43c0

File: slack-api-model-kotlin-extension/src/main/kotlin/com/slack/api/model/kotlin_extension/block/composition/dsl/RichTextObjectDsl.kt (hunk 6 above)
Comment (KENNYSOFT) [reply slack-api-model-kotlin-extension/src/main/kotlin/com/slack/api/model/kotlin_extension/block/composition/dsl/RichTextObjectDsl.kt]: Resolved in 1be897c8ce41fb2d32ee1b4cdaa0d647fd7b43c0

File: slack-api-model-kotlin-extension/src/main/kotlin/com/slack/api/model/kotlin_extension/block/composition/dsl/RichTextObjectDsl.kt (hunk 7 above)
Comment (KENNYSOFT) [reply slack-api-model-kotlin-extension/src/main/kotlin/com/slack/api/model/kotlin_extension/block/composition/dsl/RichTextObjectDsl.kt]: Resolved in 1be897c8ce41fb2d32ee1b4cdaa0d647fd7b43c0

File: slack-api-model/src/main/java/com/slack/api/model/block/element/RichTextQuoteElement.java (hunk 8 above)
Comment (KENNYSOFT) [reply slack-api-model/src/main/java/com/slack/api/model/block/element/RichTextQuoteElement.java:22]: Resolved in 3afa2dcd02e0dbaaf4420f68c09870f303dc9eb9

**Review by seratch — APPROVED** [review]
Looks good to me! Once the CI builds pass, I will merge this PR. Thank you so much for the contribution!

**Comment by seratch** [comment]
Oops, let me correct myself about the merge timing, we'll release this patch release before it: https://github.com/slackapi/java-slack-sdk/milestone/108 After that, this enhancement will be included in an upcoming minor release.
